colossalai/pipeline/policy/bert.py

```diff
@@ -0,0 +1,326 @@
+
+    #output of non-first and non-last stages:
+    if not return_dict:
+        return tuple(v for v in [
+            hidden_states,
+            next_decoder_cache,
+            all_hidden_states,
+            all_self_attentions,
+            all_cross_attentions,
+        ] if v is not None)
+
+    #return dict is not supported at this moment
+    return BaseModelOutputWithPastAndCrossAttentions(
+        last_hidden_state=hidden_states,
+        past_key_values=next_decoder_cache,
+        hidden_states=all_hidden_states,
+        attentions=all_self_attentions,
+        cross_attentions=all_cross_attentions,
+    )
+
+
+# The layer partition policy for bertmodel
+class BertModelPolicy(Policy):
+
+    def __init__(self, stage_manager: PipelineStageManager, num_layers: int, num_stages: int):
+        self.stage_manager = stage_manager
+        self.layers_per_stage = self.distribute_layers(num_layers, num_stages)
+
+    def get_hold_layers(self, module: BertModel) -> List[Module]:
+        """
+        get pipeline layers for current stage
+        """
+        hold_layers = []
+        if self.stage_manager.is_first_stage():
+            hold_layers.append(module.embeddings)
+        num_layers_per_stage_accumulated = self.convert_into_accumulated()
+        hold_layers.extend(module.encoder.layer[num_layers_per_stage_accumulated \
+                    [self.stage_manager.stage-1] if self.stage_manager.stage > 0 else 0:
+                    num_layers_per_stage_accumulated[self.stage_manager.stage]])
+
+        if self.stage_manager.is_last_stage():
+            hold_layers.append(module.pooler)
+
+        return hold_layers
+
+    def get_shared_params(self, module: BertModel) -> List[Dict[int, Tensor]]:
+        '''no shared params in bertmodel'''
+        pass
+
+    def replace_forward(self, module: Module) -> None:
+        module.model.forward = MethodType(partial(bert_model_forward, stage_manager=self.stage_manager), module.model)
+
+    def distribute_layers(self, num, stage_num) -> List[int]:
+        """
+        divide layers into stages
+        """
+        quotient = num // stage_num
+        remainder = num % stage_num
+
+        # calculate the num_layers per stage
+        layers_per_stage = [quotient] * stage_num
+
+        # deal with the rest layers
+        if remainder > 0:
+            middle_stages = (stage_num - 1) // 2
+            right_extra = remainder // 2
+            left_extra = remainder - right_extra
+
+            #divide the rest part
+            left = 0
+            right = 0
+            while left_extra > 0:
+                layers_per_stage[middle_stages - left] += 1
+                left_extra -= 1
+                left += 1
+            while right_extra > 0:
+                layers_per_stage[middle_stages + right + 1] += 1
+                right_extra -= 1
+                right += 1
+        return layers_per_stage
+
+    def convert_into_accumulated(self) -> List[int]:
```

Comment:
Maybe you can use `numpy.cumsum()` directly

Reply:
done